[WIP] Adding hs.axuielement to core

asmagill opened the pull request:
I plan to review the documentation tomorrow to make sure it's oriented appropriately now that this is going to be in core, but functionally, this is the paired down version that @latenitefilms and I have been using the last couple of weeks (and other than removing some things we *weren't* using that often (if at all) and relaxing the errors when attempting to use the shortcut functions for missing properties, what we've been using for over 2? years now), so this is probably pretty close to ready for merging.

@cmsj, since this has a name so similar to `hs.uielement` (since you've recently had success with working on it when I have a track history of breaking it, I now consider you the uielement expert 😜), do you have any suggestions on a blurb to add to the top of this (and maybe uielement as well) to distinguish them? Make clear their separate roles? Or do you think that's already clear enough in their introductions?

asmagill commented:
I'll also address stickler tomorrow as well. And eff-U Travis! You *always* fail during tests when I submit things!

cmsj commented:
I’m running a local instance of Xcode server now, but it only watched master. I’ll look for something that can trigger it to test PRs.

asmagill commented on extensions/axuielement/init.lua:
```diff
@@ -0,0 +1,362 @@
+local buildTreeHamster
+buildTreeHamster = function(self, prams, depth, withParents, seen)
+    if prams.cancel then return prams.msg end
+    coroutine.applicationYield()
```
Yes, `coroutine` is a built-in library, but we've added the `applicationYield` function... how do I tell stickler to ignore something again?

asmagill commented on extensions/axuielement/init.lua:
```diff
@@ -0,0 +1,361 @@
+        if _:match("^set%u") then
+
+             -- check attributes
+             for __, v in ipairs(objectMT.attributeNames(self) or {}) do
```
AARGGHH... I'm actually using _ so I had hoped __ would also be given a pass... is there an easy way to add __ to the "pass" list?

I'm currently rebuilding and running tests here just to make sure I didn't break anything, so I'll check the docs as well...

asmagill replied:
Hmm.. looks like `-- luacheck: ignore __` at the top of the file might do it... waiting for a compile to complete to try it. However, if I'm reading things correctly this means that *any* use or misuse of __ would be ignored, not just "unused". Wish it had finer granularity, but I guess that's ok...

asmagill commented:
@cmsj Ok this Travis fail is actually legitimate (but still, eff-U Travis! I stand by my earlier lament!)

~~~
1513 HSrequire_all
1514  testRequireAll, ((res) equal to (@"")) failed: ("type of 'hs.axuielement' is 'boolean', was expecting 'table'") is not equal to ("") - one or more require statements failed
1515  /Users/travis/build/Hammerspoon/hammerspoon/Hammerspoon Tests/HSrequire_all.m:32
1516  ```
1517    NSLog(@"%@", res);
1518    XCTAssertEqualObjects(res, @"", @"one or more require statements failed");
1519}
1520  ```
~~~

This module does check at the beginning of `init.lua` to see if accessibility is enabled and returns nil (and a message to the console) if it isn't...

What behavior would you rather see in such a case? Without accessibility and a reload of Hammerspoon (relaunch?) this module is basically useless...

asmagill commented:
Ok, I haven't fully tested actually *using* the module with accessibility disabled, but I did confirm that it can work if accessibility is enabled *after* launching Hammerspoon and loading the module, but *before* actually using anything within it so... I removed the `nil` return. Let's see if Travis smiles on me now...

asmagill commented:
Yay! Travis likes me for this nano-second!

@cmsj, however, the question remains... what should a module that determines it can't run for whatever reason return? (say if it requires a specific macOS version, etc.)

asmagill commented:
Ok, this probably should be in another pull at a later date, but I wanted to get some feedback as I look into other blocking activities that might benefit from a rewrite now that this module is going into core...

https://github.com/asmagill/hammerspoon-config-take2/blob/master/utils/getMenusIn.lua is a potential replacement for `hs.application:getMenuItems`. I've extended it to replicate both the callback and non-callback versions. It also takes an optional boolean allowing the includion of the Apple menu in its results (default false, and not set for this test). A sample run comparing times follows:

~~~lua
-- it was already loaded so I pulled the relevant function from package.loaded
> gm = package.loaded["utils.getMenusIn"].getMenuItems 

-- my Safari application currently has 12 windows open (probably not relevant, since
-- the functions start at the application menubar) and about a bajillion unsorted
-- bookmarks saved in the Bookmarks menu, which probably *is* relevant...

-- using my function with no callback
> local s = os.time() ; a = gm((hs.application("Safari"))) ; print(os.time() - s)
14

-- using the Objective-C version with no callback
> local s = os.time() ; a = (hs.application("Safari")):getMenuItems() ; print(os.time() - s)
12

-- using my version with a callback
> local s = os.time() ; gm((hs.application("Safari")), function(m) print(os.time() - s) end)
-- no spining ball or apparent slowdown/lockup of Hammerspoon console
12

-- using Objective-C with a callback
> local s = os.time() ; a = (hs.application("Safari")):getMenuItems(function(m) print(os.time() - s) end)
-- spining ball appears and Hammerspoon console locks up until done
12

-- back to my version without callback
> local s = os.time() ; a = gm((hs.application("Safari"))) ; print(os.time() - s)
10

-- back to Objective-C version without callback
> local s = os.time() ; a = (hs.application("Safari")):getMenuItems() ; print(os.time() - s)
11

-- more samples just to get more times, but it basically follows the same patterns within an
-- expected margin of error based on other mac activity that we can't control

-- remove the time it takes for hs.application to search for Safari
> safari = hs.application("Safari")

> local s = os.time() ; a = gm(safari) ; print(os.time() - s)
9

> local s = os.time() ; a = safari:getMenuItems() ; print(os.time() - s)
11

> local s = os.time() ; a = gm(safari, function(m) print(os.time() - s) end)

11

> local s = os.time() ; a = safari:getMenuItems(function(m) print(os.time() - s) end)

12

> local s = os.time() ; a = gm(safari) ; print(os.time() - s)
10

> local s = os.time() ; a = safari:getMenuItems() ; print(os.time() - s)
11
~~~

I know this isn't anything like formalized testing, but based on these results, I'm thinking a complete replacement would probably be fine with no real impact or change to anyones existing usage of the method.

It should also be noted that except for the new version with a callback, all of these resulted in a spinning ball and temporary lockup of Hammerspoon -- the Objective-C version with a callback just started the spinning ball a second later (which makes sense because it queues its work so it doesn't start until the application runloop process another "event".

I'm kind of relieved that the Objective-C and lua based version without a callback are so close in time and it actually varies as to which one is "better"... I was a little afraid that the time differences would be large enough that we'd have to consider maintaining both a compiled version and the lua version side-by-side.

Thoughts?

asmagill commented on extensions/axuielement/init.lua:
```diff
@@ -0,0 +1,509 @@
+
+    -- getters for paramaterizedAttributes
+    for __, v in ipairs(objectMT.parameterizedAttributeNames(_) or {}) do
+        if partialName then
```
ok yeah, I curse stickler, but this one was actually valid

asmagill commented on extensions/axuielement/init.lua:
```diff
@@ -0,0 +1,509 @@
+
+    -- doers for actionNames
+    for __, v in ipairs(objectMT.actionNames(_) or {}) do
+        if partialName then
```
ok yeah, I curse stickler, but this one was actually valid

asmagill commented:
I'll deal with Stickler tomorrow.

I've moved this to WIP because there are a couple of things I think I can get done in the next day or so that I'd like to have in place. No more than a couple of days and this will be ready.

I've added `hs.axuielement:elementSearch` which is kind of the power-house for collecting elements that you don't already know the direct path to. `allChildElements` and `buildTree` are now syntactic sugar for this method with specific options set. `elementSearch` requires a callback but it now uses a breadth-first search rather than depth-first, and no longer uses recursion, so, while it's still slow for applications with a lot of elements, it *is* noticeably faster -- On my Safari with 6 windows open, the original `allChildElements` took on average 300 seconds while the new one averaged closer to 190. Almost 10,000 elements were returned each time (I don't recall the exact number because it changed as I opened and closed windows and tabs, but both versions returned the same numbers when I let it sit idle in the background long enough)

@latenitefilms This does slightly change the results for `buildTree` -- now if the depth is exceeded, the attribute value at that level will be the unexpanded axuielement rather than the string "** maxDepth reached" (that's close, I forget the exact text). Otherwise, it should be functionally identical.

I've also fixed a couple of example files:
* https://github.com/asmagill/hs._asm.axuielement/blob/master/examples/dockStuff.lua which shows some things that can be done with the Dock.
* https://github.com/asmagill/hs._asm.axuielement/blob/master/examples/axtypeMarker.lua which draws rectangles around elements of the specified type for an application or window.

asmagill commented:
Added AXTextMarker and AXTextMarkerRange support... at least I *think* so. If you recall when I first started working on this, I found one reference in the AXUIElement documentation that says these are possible attribute value types but never found further documentation or a reliable way to get at the data in them... I finally located their functions by digging through some public chrome library sources (the link is in the comments of `common.h`) and back-checking it against Hopper's disassembly of the HIServices framework. More importantly, it doesn't crash and *does* give valid results when the types are encountered in Safari attributes.

But I don't understand the binary data in the `bytes` field (see https://github.com/asmagill/hs._asm.axuielement/blob/master/README.md for an example).

I'm debating whether core should contain the code to `set` them until we know more about the format of the binary data which I hope further digging into the chrome sources might elucidate, but for now, I figure since you won't find the properties unless you're experimenting or specifically looking for them, it's probably ok... unless someone disagrees?

It's also possible, if I (or someone else if they wish to contribute) *can* untangle what the binary data means, that it may be easier to use if the data format changes (possibly to a userdata of its own with appropriate methods), so it should be considered experimental at this stage. But at least now we shouldn't see any more "unrecognized value type (###)" messages for property values.

cmsj commented:
> @cmsj, however, the question remains... what should a module that determines it can't run for whatever reason return? (say if it requires a specific macOS version, etc.)

I think the module should print an error to the Console and return `nil`.

latenitefilms commented:
@asmagill - As a wild guess, could the `bytes` just be the binary `NSString` value of the selected portion of text?

asmagill commented:
Tracking down some functions within the chromium library I mentioned above, I found a comment in a function used to create the object they use to initialize the AXTextMarker objects:

~~~
// A serialization of a position as POD. Not for sharing on disk or sharing
// across thread or process boundaries, just for passing a position to an
// API that works with positions as opaque objects.
~~~

If they are truly opaque, then they only have meaning within the app itself, so their only use would be if they were expected to be the values passed in when querying at least some of the parameterized attribute values, but I haven't had a chance to try this yet. This would make sense (though it would be nice to find something from Apple, or anyone really, confirming or refuting this) since we *can* just treat them as a CFTypeRef to pass them to `AXUIElementCopyParameterizedAttributeValue` without ever having to know what they mean or contain, and hence the lack of any real documentation about them.

But I probably won't be able to test this theory until tomorrow or Monday.

asmagill commented:
Ok, I can confirm that the text markers do work as arguments to at least *some* of the parameterized attributes.

Parameterized attributes are kind of a PITA... there is no easy way that I've found to figure out what the parameter type a give parameterized attribute expects except by trial and error... I've found a handful documented *as suggestions* in comment strings for framework headers, but not many... I'm in the process of collecting these into a form I can include in the documentation, not sure exactly where yet, but I think they will always be something of a challenge when determining their usefulness with a given application.

I'm going to shoot for having this ready to merge (with documentation about what little I have found out about parameterized attributes and a couple of bug fixes I haven't pushed yet) by the weekend. After that, I hope to get `hs.text` and `hs.userpreferences` moving the following week. `hs.text` is probably close to ready as it is, but I'd really like to get a lua-string regex submodule for it finished as well. And as to `hs.userpreferences`, that will probably take a little longer as I need to go through earlier post and see how to best organize it since it will be pulling pieces from my cfpreferences, the existing `hs.plist` and CommandPost's `cp` library.

latenitefilms commented:
AMAZING!!

I guess also, you can always merge in these new extensions once they're functional and documented (I feel like `hs.text` is pretty good as is?), and add new features (such as a lua-string regex submodule) when they become ready.

In regards to `hs.userpreferences`, you could just merge in a version that simply offers `:get()`, `:set()` and `:sync()` so that `hs.donotdisturb` can be finished - then tackle some of the other more complicated issues (such as `NSKeyedArchiver` support) later down the line?

I feel like once the community start playing with and using `hs.axuielement`, `hs.text`, etc. there might be even more people offering ideas/comments/suggestions for situations we haven't even thought of - which is why I'm kinda keen to merge all these extensions sooner rather than later.

But of course, as always, no pressure or rush. I HUGELY appreciate all the amazing work you're doing for the Hammerspoon (and CommandPost!) communities!!

Stay safe!

asmagill commented:
Ok, for those still playing the home game, the version at https://github.com/asmagill/hs._asm.axuielement 
* has an updated `elementSearch` that now takes a `count` attribute to trigger the callback early and a `next` method to resume the search where it left off.
* moves AXTextMarker and AXTextMarkerRange into userdata objects to make them a little less awkward to use as parameters to parameterized attributes (the `newRange` constructor also gives an example of their use)

I still want to see if I can come up with a few more details about parameterized attributes to include in the core documentation and test the changes to `elementSearch` a little more, but I think I'll probably land this tomorrow if I don't run into any issues. Beyond a few basics that I would like to include in the internal docs, I have some plans for a couple of wiki pages, but those can be added at any time and this needs to land, once I'm sure it won't be changed or added to anytime soon.

asmagill commented:
@cmsj, @latenitefilms, got a question on style regarding specifying text ranges for parameterized attributes that I'd like some input on... I know neither of you have used these before, but this is more about presentation...

I've almost completed the documentation on the "official" parameterized attributes found in the public header files (and still finding obscure bugs that might have otherwise gone unnoticed if I weren't poking around so much to test things out, so current target is to merge this Friday).

Not surprisingly, a lot have to do with text in apps like Terminal, TextEdit, etc.

As an example, you can do something like:

~~~lua
termWindow = hs.axuielement.applicationElement(hs.application("Terminal"))("AXWindows")[1]("AXSections")[1].SectionObject

termWindow("AXStringForRange", { location = 10, length = 10 })
~~~

And get the characters in the window history as a string corresponding to the range specified.

Now, the keys `location` and `length` were chosen because they match the definition of a `CFRange`, which is the value this parameterized attribute requires.

As you'd expect since its based on macOS data types, it's `0` based (the first character is at location 0). For the sake of being closer to the Lua style, I've also allowed the specifying of ranges (from Lua to CFRange) as `{ starts = #, ends = # }`, which is `1` based (so the above could have used `{ starts = 11, ends = 20 }` instead), but because we usually don't know how large the full content is negative indices aren't supported (some elements include content length in another attribute, others don't, and the parser is recursive to handle tables, so it usually doesn't know what element we're working with anyways).

Now consider Safari (and likely anything based indirectly on WebKit like the `WKWebView` class) which uses `AXTextMarker` and `AXTextMarkerRange`. While we can't know precisely what a given `AXTextMarker` means location wise because it seems to be application specific, it's interesting to note that `AXTextMarkerRangeCreate` takes a starting `AXTextMarker` and an ending `AXTextMarker` as arguments.

So the question is should I forgo the `{ location = #, length = # }` format all together and just have CFType conversion be to/from `{ starts = #, ends = # }`?

Pros:
* similar to lua's handling of indices for strings, so no forgetting to add/subtract that pesky 1 when manipulating the results with the Lua string library.
* similar to constructor for `hs.axuielement.axtextmarkerrange` objects (and the underlying `AXTextMarkerRange` CFType)

Cons:
* `1` based if we want true Lua similarity... 
  * this is in contrast to the other text oriented parameterized attributes which take or return a character index position or a line number
  * since these are just integers, we can't have the CFType conversion function automatically add 1 because it never knows where the type it's converting came from, just the type
  * and even if we somehow managed to track this and used a lookup table to see if the attribute matches some criteria, everything documented is just a "suggestion" by Apple -- application developers are free to add their own attributes as they see fit
* Not a true representation of the data being returned -- it makes a kind of sense when `CFRange` corresponds to text, but may or may not for other contexts which return/take `CFRange`

Actually now that I've laid it all out like this, I'm leaning towards adding a conversion function to `hs.axuielement`'s init.lua that allows you to convert a returned `CFRange` to the starts/ends format if you really want that format, leave the internal  handling as it is, and making sure to document this distinction in the documentation and future Wiki pages about `hs.axuielement`. 

Unless this has sparked specific thoughts or preferences?

cmsj commented:
I like the sound of surfacing Apple’s data directly, with some conversion functions.

latenitefilms commented:
I'm happy to defer to whatever you two think is best!

asmagill commented:
I've redone elementSearch and matchesCriteria which will make the wiki pages I have planned easier, but the core functionality remains the same.

I'll deal with stickler while running through one last check of bugs and documentation review tomorrow.

asmagill commented:
@latenitefilms , @cmsj, style question (since this is potentially a breaking change, I want to make a decision before landing this):

Ok, using metamethods, you can access an accessibility object's attributes and children in a couple of different ways that are a little simpler than using the formal methods (for clarity, I'm using the formal name of attributes in the examples, but it should be noted that there is logic included that allows you to use an "informal" version of attribute names in many places... e.g. if "AXTitle" is a valid attribute for an element, you can refer to it as "title" or "Title" or "AXTitle" with the psuedo functions created by the metamethods, though not when using the formal methods -- `object("Title")` works but `object:attributeValue("Title")` will return `nil,	"Attribute is not supported by target."`)

To access children, `__len` returns the number of direct children (i.e. the number of members in the "AXChildren" attribute), and `__index` treats integer keys as indices into the "AXChildren" array. So you can use things like `if #object > 0 then ...` and `for i,v in ipairs(object) do ...` to access an object's children.

To access attributes of the element, `__index` with string keys returns a function to get the attribute named (assuming it doesn't match a formal method of the module) and `__call` uses these functions as well... `object("AXTitle")` and `object:AXTitle()` are equivalent to `object:attributeValue("AXTitle")`.

It's this latter I'm not so sure about... It seems to me that while the `__call` approach is fine (and in fact is what I generally use myself), it would make more sense for the `__index` function to return the value of the attribute rather then a function to get the value (i.e. `object.AXTitle` should return the value of the attribute rather than `object:AXTitle()`). This would also have the benefit that `object.AXTitle` could return nil when the attribute doesn't exist rather than the error that `object:AXTitle()` returns because there is no pseudo function named `AXTitle` for that particular element. `__newindex` could also be added to allow setting attributes which can be changed with `object.AXTitle = X` (and generate an error if the attribute isn't settable)

It would then make sense for `__paris` to iterate over the attributes so `pairs` could be used to visit them all...

* `for i,v in ipairs(object) do ...` iterates over children (this is currently implemented)
* `for k,v in pairs(object) do ...` iterates over attributes (this would be a change)

Currently `__paris` returns the psuedo functions, but I think this new approach would be more useful and intuitive (you can always use `object: dynamicMethods()` if you truly want the psuedo functions, which honestly, I've only ever used internally within the module itself and not for anything else).

I would probably leave the `__call` method as is (though it will need a few changes internally) because it's also used to allow things like `object("doAXPress")` to perform the "AXPress" action (equivalent to `object:performAction("AXPress")`... or should I limit `__call` to just actions and parameterizedAttributes? That actually seems a little cleaner and maybe a little easier to implement, though I'll have to make changes to some of the examples...

Thoughts? I know this delays this another day or so, but I want to lock in the syntax so that going forward changes are only additions/bug-fixes rather than be stuck with a confusing and hard to maintain/debug mish-mosh in the name of maintaining backwards compatibility...

latenitefilms commented:
Thoughts @randomeizer ?

asmagill commented:
@randomeizer (sorry, I forget you do a lot with this module as well... @latenitefilms is just more vocal I guess 😜) - Apparently the internally used function for setting a value, `AXUIElementSetAttributeValue` doesn't return an error if you try to set an attribute that isn't settable as identified with `AXUIElementIsAttributeSettable`... it just silently fails to make any change... should this module follow suit (i.e. go ahead and make the attempt even though we know it will do nothing)? Or would you rather an error so you know you're attempting to do something that will have no effect?

randomeizer commented:
I’d have to review our code, but in general we have been using the `attributeValue`/`setAttributeValue` methods because they didn’t just throw errors if referencing something that doesn’t exist, so not much would break by having the `element.attribute` call return and/or set the value directly, rather than via a function. We’ve also been refactoring a lot of our code to prefer using properties rather than functions for navigating around object hierarchies these days, since it reduces the `()` stutter when working with hierarchical models.

I wasn’t even aware that the `__call` method was an option. Feels kind of weird to me personally, although I guess it’s just a shortcut for calling `:attributeValue(...)` or `:setAttributeValue(...)`?

So, my vote would be:

1. Have the convenience calls via `__index` return or set the value directly, rather than through a function.
2. Don’t throw error messages when getting or setting an attribute that doesn’t exist (as long as there is some other way to explicitly test that an attribute is actually supported on a given AXUIElement).
3. `ipairs` returns the list of `AXChildren` (as it does currently) and `pairs` returns the values, not the pseudo functions.

Now that I’ve said we don’t use the psuedo functions currently, no doubt a whole bunch will show up, but in general, I agree: worth the pain of getting the API right before adding it into the mainline.

asmagill commented:
I'v been experimenting with what changes this would require, and I've found one issue that differs from what you've outlined -- the case of trying to *set* an attribute that doesn't exist for the element at all...

Using `hs.timer` as a test for what "normal" behavior of a userdata object is, I observe:

~~~lua
> t = hs.timer.doEvery(1, function() print("in timer") end):stop()

> t.abc
nil

> t.abc = "2"
[string "t.abc = "2""]:1: attempt to index a hs.timer value (global 't')
~~~

Now since we're mimicking table like behavior with these metamethods, the case can be made that the "setter" should just silently fail... but then the argument can be made that for a true table, it *wouldn't* fail, so there is no expected behavior (as far as I've been able to determine, the only value that is not allowed to be a table key is `nil` -- even booleans and reserved command words are acceptable, though you have to set them like `{ [true] = 1, [false] = 0, ["not"] = -1 }` -- only `{ [nil] = 0 }` throws an error, thus far in my tests)

So... should `__newindex` silently fail for non-existent attribute names? Personally I lean towards no, as in my mind there is a difference between *not* changing a value of an existing attribute and trying to manipulate something that doesn't exist -- generally I'd like to know when the latter is occurring as it's probably a typo on my part. But I'd like your thoughts on the matter.

- - -

Also, I don't know if you've tracked the latest changes, now most of the official methods return nil and an error string as the second return value (as well as logging the error) when an error occurs... for example:

~~~lua
local ok, msg = hs.axuielement.applicationElement("Hammerspoon"):setAttributeValue("nonexistentAttribute", true)
if not ok then print(msg) end
~~~

will print "Attribute is not supported by target."

(It should be noted that if "nonexistentAttribute" were instead "AXTitle", no error would be returned, even though `hs.axuielement.applicationElement("Hammerspoon"):isAttributeSettable("AXTitle")` returns false... this actually surprised me, but it does simplify the code a little bit to conform to what you've outlined as preferred above)

This should be transparent to most existing code since the first return value hasn't changed.

asmagill commented:
An interesting find... `AXFrame` appears to be a recognized accessibility attribute for windows (at least in the 5 or 6 apps I just tested) that you can get a value for, but outside of `hs.axuielement:attributeNames` and `hs.axuielement:attributeValue`, everything else claims "Attribute is not supported by target."... (It should be noted that "AXFrame" is *not* one of those included in the support tables corresponding to an official accessibility properties, though honestly, I'm not sure how up to date Apple keeps that list... I only noted a handful of changes between when I first created those tables (at least three years ago) and when I last reviewed them a couple of days ago.)

Thus far in my observations, elements that have an AXFrame, also have an AXPosition and an AXSize, both of which are settable... should a special case be made when trying to set this value so that it instead sets position and size behind the scenes for you? Or should I just make note of it somewhere in the docs?

asmagill commented:
Ok, I've simplified the dynamic methods a bit, removing the __call method approach entirely. I want to run a few more tests before updating this pull. You can read more about the changes at https://github.com/asmagill/hs._asm.axuielement/blob/refactorMetamethods/Reference_Core.md but I give a synopsis below.

The main changes:
* __call is now undefined. This reduces extra parenthesis and quotes making things easier to read.
* the shortcuts now require the attribute to be specified with its real name (you can no longer leave off AX). This makes it less likely that there will be a name collision with a formal method, and makes it more obvious when referring to the same attribute -- you don't have to remember that `.title`, `.Title`, and `.AXTitle` *might* refer to the same attribute. It also simplifies some of the internal logic.

Basic shorthands:
* `obj.AXAttribute` will return the value of the attribute `AXAttribute`. If `AXAttribute` isn't a valid attribute for the element, returns nil
* `obj.AXAttribute = ...` will attempt to set the value of attribute `AXAttribute`. See the docs for why I say "attempt". If `AXAttribute` isn't a valid attribute for the element, causes an error (just like assigning a value to a field for any userdata object would).
* `obj:doAXAction()` will perform `AXAction` on the element. If `AXAction` isn't a valid action for the element, causes an error (just like calling a non-existent method on any userdata object would).
* `obj:AXAttributeWithParameter(value)` will return the value for the parameterized attribute `AXAttribute` when requested with the specified value. If `AXAttribute` isn't a valid parameterized attribute for the element, causes an error (just like calling a non-existent method on any userdata object would).

@randomeizer what are your thoughts on the three cases above that *can* throw errors? 

My reasoning for doing so is as follows:
* When setting a property value, I think a typo is the most likely reason for the wrong attribute to be specified and I'd want to know as early as possible (when getting an attribute, the `nil` value will probably cause an issue soon enough, unless I know that it's a possibility and code for it, which is why that one doesn't error out).
* For actions and parameterized attributes, making these *not* throw an error means that there would need to be a wrapper for *any* string that started with "do" or ended with "WithParameter"... there is no definitive list of valid actions or parameterized attributes to limit it to and app developers (and Apple) can (and do) add them as they wish, so we can either limit it to what the element specifically supports, or dynamically make an empty function for *all* of them that match the format.

In the end, I will be but one user of the module, so if you really think it will be easier for any of the three cases above to just silently fail to do anything, I can change them, but I'm not a fan of processes that give no easy way to detect failures (especially typos or attempting to work with the wrong object) near where they actually occur.

randomeizer commented:
I'm generally in favour of simplification, so I think the naming choice makes sense. There may be a couple of attributes that we call as functions (eg `children()`, perhaps) that will now have to be `AXChildren`, but that will be changing regardless, so no big deal there, and we really don't use them often. Probably the only "breaking" change would be the `doAXAction()` change, since I think there are a few places where we call `doPress()`, but they shouldn't be hard to find. I prefer consistency over workarounds for backwards-compatibility.

And having an error thrown when doing something that should have an effect but doesn't (setting/'do'ing, etc) seems reasonable.

Regarding the `AXFrame` question, not sure. I could be persuaded either way. Being able to set `AXFrame` kind of makes sense, but I'm guessing in practice I would probably set them independently anyway? Maybe just stick with the built-in functionality: keep it read-only if appropriate, and document it.

asmagill commented:
Ok, for core functionality, this is ready, but I do have a couple of questions...

First, I've gone through the documentation and I think I've cleared up all of the places where I used the term "child" or "children" but really meant "descendants"... Children of AXUIElement's are those specifically belonging to the AXChildren attribute and this is reflected with the `__len` and `__index` (when the key is an integer) metamethods. Descendants are those elements which are children of an element, or children of children of an element, or ... basically any element (except AXParent and AXTopLevelUIElement) which can trace its parentage path back through the element we're starting with.

But there is one place where I haven't made the change because it will potentially be a breaking change if someone is using the method:  `hs.axuielement:allChildElements(...)`... though I've updated the documentation for the method, a more accurate method name would be `allDescendants()` or `allDescendantElements()`.  Objections if I change it to one of these before landing this?

Secondly, I'm generally of the mind that modules should provide basic functional building blocks and let others develop their own library functions on top of them that encapsulate reusable code for use within their own implementations. My ideas of how something can be used is often limited and I have been surprised at times to see what others have done with modules I've introduced -- contrariwise, with some of the modules which try to hide a lot of the internals, I've often found myself grumbling that I wasn't able to access more fundamental methods to do something the original author hadn't thought of without creating a new pull request and adding functionality.

Where I'm going with this is that I've had a chance to review CommandPost's https://github.com/CommandPost/CommandPost/blob/develop/src/extensions/cp/ui/axutils.lua and some of these look really useful and especially helpful for someone new to accessibility objects and how they're usually organized within an application. We don't have a really good place for helper libraries like this... modules for fundamentals, Spoons for "completed" works, but nothing really in between that's entirely lua code, but not complete enough by itself to be a Spoon.

To get this module moving, I'm going to work on integrating some of these tonight into the module, but I thought I'd put my thoughts out there... do we need something in between modules and Spoons for "helper" libraries? From the developer perspective, I lean towards yes, but wonder if it might just add confusion for newer non-programmer users?

randomeizer commented:
We don't use `allChildElements()`, so no problem there. Is `children()` still a function? I know it was only around because of the automatic attribute conversion, but we use it a fair bit.

And yes, no problem with borrowing functions from `axutils` (or anything else). Regarding modules vs Spoons, not sure. We don't really use Spoons in CP, since we have our own plugin system. In general, we put 'frameworks' in modules and managers in plugins, but it's not a hard line.

In the case of `axutils` specifically, most of the functions there would work just as well (or better) integrated directly into the `axuielement` class. Some functions are more pure "functions" (such as the comparison functions), so not sure the best place to put those.

Not sure if there is a good 'in between', other than what we've done here: create a module with some functions.

asmagill commented:
To be fair, the question of "where things go" is a separate issue from this pull, it's just one that kind of finally crystallized for me as I considered how best to integrate some of axutils functions into the module.

As to `:children()`, since it is a reference to the `AXChildren` attribute, the syntax *has* changed with the latest in this pull and at https://github.com/asmagill/hs._asm.axuielement in two ways:

* attributes are now considered key-value pairs for the userdata object, e.g. `obj.AXChildren` rather than functions -- this allows the `__index` metamethod to just return `nil` for a non-existent attribute rather than try to determine if it needs to return a function which returns `nil` for every key that doesn't exist as an attribute name

* attributes now require their full formal name, e.g. `obj.AXChildren`, rather than `obj.children` -- it simplifies some of the internal logic and reduces the confusion that can occur when you have to realize that different labels can refer to the same thing,

Thinking about it, though, getting an attribute's children *is* a common enough thing that a case could be made for having an explicit method that returns an array of the children, or an empty table if the element has no direct children... on the other hand you can also do `for i,v in ipairs(obj) do ... end` to iterate through the children, so it's not "necessary", but it might be clearer?

*(note that if you access the AXChildren property, it *could* be nil, so you'd need to do `for i,v in ipairs(obj.AXChildren or {}) do .. end` to protect against it -- accessing the userdata itself as an array with `ipairs` doesn't have this limitation)*

So short version, yes, that will need to change in your code unless the consensus is that it's a common enough pattern that an explicit method should be added. Thoughts?

randomeizer commented:
Yeah, I’m not sure with `children()`. It’s already getting treated as a special case (via `ipairs`), so not sure if it needs anything further. If you’ve looked at the `axutils.children(...)` function, you’ll note that it basically adds support for checking `children` and if it doesn’t exist then looping directly on the element. This also lets the same function work on ‘list’ tables of elements as well. Not suggesting that it gets integrated into `axuielement` (I don’t think it really applies), just that we will probably adapt that function to work with whatever the final API becomes.

cmsj commented:
> Where I'm going with this is that I've had a chance to review CommandPost's https://github.com/CommandPost/CommandPost/blob/develop/src/extensions/cp/ui/axutils.lua and some of these look really useful and especially helpful for someone new to accessibility objects and how they're usually organized within an application. We don't have a really good place for helper libraries like this... modules for fundamentals, Spoons for "completed" works, but nothing really in between that's entirely lua code, but not complete enough by itself to be a Spoon.
> 
> To get this module moving, I'm going to work on integrating some of these tonight into the module, but I thought I'd put my thoughts out there... do we need something in between modules and Spoons for "helper" libraries? From the developer perspective, I lean towards yes, but wonder if it might just add confusion for newer non-programmer users?

I think this is a really good point, and rather than have the discussion here, I've opened #2405.

asmagill commented:
@cmsj, just saw that Lua 5.4 has been officially released; it has taken me a bit longer to get time to evaluate and test the additions I want to add to this, but what's currently here, I consider ready... if you want to get a release out in the next few days so you can start integrating Lua 5.4 into Hammerspoon, this pull can land as is and I can hold off on my additions and do another pull after the release.

Otherwise, I hope to have time to get back to this later this week or weekend.

cmsj commented:
@asmagill I can't do a release until I've addressed #2360 , unfortunately.

cmsj commented:
#2407 has the specific hashes, but `master` now uses Lua-5.4.0.

latenitefilms commented:
FYI @asmagill: the Lua 5.4 update broke `hs._asm.axuielement`:

```lua
2020-07-01 01:10:27: -- Loading /Users/chrishocking/Github/CommandPost/src/extensions/cp/init.lua
2020-07-01 01:10:28: *** ERROR: error loading module 'hs._asm.axuielement.internal' from file '/Users/chrishocking/Github/CommandPost/src/extensions/hs/_asm/axuielement/internal.so':
	dlopen(/Users/chrishocking/Github/CommandPost/src/extensions/hs/_asm/axuielement/internal.so, 6): Symbol not found: _lua_newuserdata
  Referenced from: /Users/chrishocking/Github/CommandPost/src/extensions/hs/_asm/axuielement/internal.so
  Expected in: /Users/chrishocking/Github/CommandPost-App/build/CommandPost.app/Contents/MacOS/../Frameworks/LuaSkin.framework/Versions/A/LuaSkin
 in /Users/chrishocking/Github/CommandPost/src/extensions/hs/_asm/axuielement/internal.so
stack traceback:
	[C]: in ?
	[C]: in function 'rawrequire'
	...Github/CommandPost/src/extensions/hs/_coresetup/init.lua:692: in function 'require'
	.../CommandPost/src/extensions/hs/_asm/axuielement/init.lua:32: in main chunk
	[C]: in function 'rawrequire'
	...Github/CommandPost/src/extensions/hs/_coresetup/init.lua:692: in function 'require'
	...hrishocking/Github/CommandPost/src/extensions/cp/app.lua:20: in main chunk
	[C]: in function 'rawrequire'
	...Github/CommandPost/src/extensions/hs/_coresetup/init.lua:692: in function 'require'
	.../CommandPost/src/extensions/cp/apple/finalcutpro/app.lua:10: in main chunk
	[C]: in function 'rawrequire'
	...Github/CommandPost/src/extensions/hs/_coresetup/init.lua:692: in function 'require'
	...CommandPost/src/extensions/cp/apple/finalcutpro/init.lua:91: in main chunk
	[C]: in function 'rawrequire'
	...Github/CommandPost/src/extensions/hs/_coresetup/init.lua:692: in function 'require'
	...rishocking/Github/CommandPost/src/extensions/cp/init.lua:23: in main chunk
	[C]: in function 'xpcall'
	...Github/CommandPost/src/extensions/hs/_coresetup/init.lua:794: in function 'hs._coresetup.setup'
	(...tail calls...)
```

See: https://www.lua.org/manual/5.4/manual.html#8.3

latenitefilms commented:
> For compatibility, the old names still work as macros assuming one single user value. Note, however, that userdata with zero user values are more efficient memory-wise.

Ah, ok... so it seems we need to recompile all the extensions with Lua 5.4 for this macro to work properly.

However, there are some lines in `hs.axuielement` that don't seem to work on Mojave:

```objc
[skin pushNSObject:(__bridge NSString *)kAXDecorativeSubrole] ;              lua_setfield(L, -2, "decorative") ;
```

```objc
[skin pushNSObject:NSAccessibilityFocusRingManipulationParameterizedAttribute] ;              lua_setfield(L, -2, "focusRingManipulation") ;
```

Commenting these two lines seems to fix things.

@asmagill - What are the chances of you pushing out new Lua 5.4 releases of:

- `hs._asm.axuielement`
- `hs._asm.cfpreferences`
- `hs._asm.undocumented.touchbar`
- `hs._asm.undocumented.touchdevice`
- `hs._asm.xml`

FYI: We also have these bundled in CommandPost, but only use them for testing:

- `hs._asm.guitk`
- `hs._asm.iokit`

asmagill commented:
Ah, I was figuring we'd have one more 5.3 release and wanted to land this in it, since it's been a while, and *then* work on 5.4, etc... but I guess not 😀

I'll see about recompiling my release bundles for modules starting tonight... I think after I upgrade to lua 5.4 on my machine and recompile Hammerspoon, it should mostly be as simple as recompiling the zip bundles, so it shouldn't take that long.

I had read that a couple of the C API functions had been replaced, but that macros mapped to the previous versions... unfortunately macros don't compile as the original function, so a recompile *is* required, even if code changes aren't.

As to the compile issues in Mojave, I did recently run through the header files to see if there were any new roles, subroles, etc. to add to the constants tables for reference, and I do recall a few new additions... I'll wrap to check at compile time for their existence -- they are just included for reference, after all, and don't affect functionality, so you commenting them out for now won't cause any issues.

latenitefilms commented:
Legend, thanks! Of course, no rush or pressure on any of this - I can recompile myself on my end for my purposes - just more a general note so you can eventually update your repos for the wider community.

FYI: We’ll probably stick with the older implementation of `hs._asm.axuielement` for now anyway until `hs.axuielement` lands in Hammerspoon’s core.

cmsj commented:
@asmagill oh, sorry, I misunderstood what you said earlier about Lua then. TBH I don't have a lot of time at the moment, what with the home schooling I'm doing, so I doubt 0.9.79 will be tremendously soon.
